batch2/batch/driver.py

```diff
@@ -0,0 +1,470 @@
+        return f'{(self.status, self.message)}'
+
+
+class Pod:
+    @staticmethod
+    def from_record(driver, record):
+        spec = json.loads(record['spec'])
+        status = json.loads(record['status']) if record['status'] else None
+
+        inst = driver.inst_pool.token_inst.get(record['instance'])
+
+        pod = Pod(
+            driver=driver,
+            name=record['name'],
+            spec=spec,
+            output_directory=record['output_directory'],
+            cores=record['cores'],
+            status=status,
+            instance=inst
+        )
+
+        if inst:
+            inst.schedule(pod)
+
+        return pod
+
+    @staticmethod
+    async def create_pod(driver, name, spec, output_directory):
+        container_cpu_requests = [container['resources']['requests']['cpu'] for container in spec['spec']['containers']]
+        container_cores = [parse_cpu(cpu) for cpu in container_cpu_requests]
+        if any([cores is None for cores in container_cores]):
+            raise Exception(f'invalid value(s) for cpu: '
+                            f'{[cpu for cpu, cores in zip(container_cpu_requests, container_cores) if cores is None]}')
+        cores = max(container_cores)
+
+        await db.pods.new_record(name=name, spec=json.dumps(spec), output_directory=output_directory,
+                                 cores=cores, instance=None)
+
+        return Pod(driver, name, spec, output_directory, cores)
+
+    def __init__(self, driver, name, spec, output_directory, cores, instance=None, on_ready=False, status=None):
+        self.driver = driver
+        self.name = name
+        self.spec = spec
+        self.output_directory = output_directory
+        self.cores = cores
+        self.instance = instance
+        self.on_ready = on_ready
+        self._status = status
+        self.deleted = False
+
+        loop = asyncio.get_event_loop()
+        self.lock = asyncio.Lock(loop=loop)
+
+    async def config(self):
+        future_secrets = []
+        secret_names = []
+
+        for volume in self.spec['spec']['volumes']:
+            if volume['secret'] is not None:
+                name = volume['secret']['secret_name']
+                secret_names.append(name)
+                future_secrets.append(self.driver.k8s.read_secret(name))
+        results = await asyncio.gather(*future_secrets)
+
+        secrets = {}
+        for name, (secret, err) in zip(secret_names, results):
+            if err is not None:
+                traceback.print_tb(err.__traceback__)
+                log.info(f'could not get secret {name} due to {err}')
+            secrets[name] = secret.data if not err else None
+
+        return {
+            'spec': self.spec,
+            'secrets': secrets,
+            'output_directory': self.output_directory
+        }
+
+    async def mark_complete(self, status):
+        self._status = status
+        asyncio.ensure_future(db.pods.update_record(self.name, status=json.dumps(status)))
```

Comment:
this seems dangerous? What if we mark_complete but the database never gets updated. Now the pod is on the complete_queue but with no status in the database.